pkg/controller/dataimportcron-controller.go

```diff
@@ -0,0 +1,611 @@
+	"github.com/pkg/errors"
+
+	batchv1 "k8s.io/api/batch/v1"
+	v1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"kubevirt.io/containerized-data-importer/pkg/common"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+// DataImportCronReconciler members
+type DataImportCronReconciler struct {
+	client         client.Client
+	uncachedClient client.Client
+	recorder       record.EventRecorder
+	scheme         *runtime.Scheme
+	log            logr.Logger
+	image          string
+	pullPolicy     string
+	cdiNamespace   string
+}
+
+const (
+	// AnnSourceDesiredDigest is the digest of the pending updated image
+	AnnSourceDesiredDigest = AnnAPIGroup + "/storage.import.sourceDesiredDigest"
+	// AnnNextCronTime is the next time stamp which satisfies the cron expression
+	AnnNextCronTime = AnnAPIGroup + "/storage.import.nextCronTime"
+
+	// dataImportCronFinalizer ensures CronJob is deleted when DataImportCron is deleted, as there is no cross-namespace OwnerReference
+	dataImportCronFinalizer = "cdi.kubevirt.io/dataImportCronFinalizer"
+
+	dataImportControllerName    = "dataimportcron-controller"
+	labelDataImportCronName     = "dataimportcron-name"
+	digestPrefix                = "sha256:"
+	digestDvNameSuffixLength    = 12
+	cronJobUIDSuffixLength      = 8
+	defaultImportsToKeepPerCron = 3
+)
+
+// Reconcile loop for DataImportCronReconciler
+func (r *DataImportCronReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	dataImportCron := &cdiv1.DataImportCron{}
+	if err := r.client.Get(ctx, req.NamespacedName, dataImportCron); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	if dataImportCron.DeletionTimestamp != nil {
+		err := r.cleanup(ctx, dataImportCron)
+		return reconcile.Result{}, err
+	}
+	if err := r.initCron(ctx, dataImportCron); err != nil {
+		return reconcile.Result{}, err
+	}
+	return r.update(ctx, dataImportCron)
+}
+
+func (r *DataImportCronReconciler) initCron(ctx context.Context, dataImportCron *cdiv1.DataImportCron) error {
+	if dataImportCron.Annotations == nil {
+		dataImportCron.Annotations = make(map[string]string)
+	}
+	if len(dataImportCron.Status.CurrentImportDataVolumeNames) == 0 {
+		dataImportCron.Status.CurrentImportDataVolumeNames = make([]string, 1)
```

Comment:
my concern with this is that it will create a list with empty string.  Why not just wait until we have a dv name and assign it like this `dataImportCron.Status.CurrentImportDataVolumeName = []string{dvName}`

Then when that dv succeeds set to nil